### ISSUE TITLE
Lightweight generics support

### ISSUE THREAD

**Issue by Manishearth** [rust-diplomat/diplomat#753]
In theory, Diplomat _could_ support generics, provided each generic can only pull from a closed set of types.

Something like this:

```rust
#[diplomat::generictrait]
pub trait SomeTrait {...}

impl SomeTrait for crate::module::ffi::SomeOpaqueType;
impl SomeTrait for crate::module::ffi::SomeOtherOpaqueType;
impl SomeTrait for u8;

pub struct GenericOpaque<T: SomeTrait>(RustType<T>);

impl GenericOpaque {
   fn create(t: &T) -> Box<Self> {...}
}
```

This would generate C ABI functions `create_SomeOpaqueType`, `create_SomeOpaqueType`, etc. In C++ this would produce a specialized, templated `GenericOpaque<T>` type. In JS it would be a single type with matchers on its methods.

If you want to use traits from other modules you'd need to define some type of `#[diplomat::trait_use = src/traits.rs]` thing so the  macro can use them (see https://github.com/rust-diplomat/diplomat/issues/752 for a similar design issue).

This is likely to be quite complicated, needing major changes to the AST, HIR, and individual backends.

**Comment by tyler-zeromatter** [comment]
I'm thinking most of the work can be shunted to the AST, since I believe generics should just be parsed in the AST and read as distinct types in the HIR:
```rs
#[diplomat::generictrait]
pub trait SomeTrait {...}

impl SomeTrait for crate::module::ffi::SomeOpaqueType;
impl SomeTrait for crate::module::ffi::SomeOtherOpaqueType;
impl SomeTrait for u8;

pub struct GenericOpaque<T: SomeTrait>(RustType<T>);

impl GenericOpaque {
   fn create(t: &T) -> Box<Self> {...}
}
```

In the AST, wherever we support generics, we can expand to account for all possibilities. Then in the HIR, we'd read something like:

```rs
pub struct GenericOpaque_SomeOpaqueType(...);

impl GenericOpaque_SomeOpaqueType {
  fn create(t: &SomeOpaqueType) -> Box<Self> {...}
}

pub struct GenericOpaque_u8(...);
impl GenericOpaque_u8 {
  fn create(t: &u8) -> Box<Self> {...}
}

pub struct GenericOpaque_SomeOtherOpaqueType(...);

impl GenericOpaque_SomeOtherOpaqueType {
  fn create(t: &SomeOtherOpaquetype) -> Box<Self> {...}
}
```

Note that this creates an exponential explosion the more generic types you add to a given struct/method, but I don't think there's a way around that.

The only main issue is with type and method names, mostly in languages that support overloading. For languages without it, we could just create those types above (maybe even let Diplomat users customize how we rename the generics? Like a format string: `#[diplomat::generic_rename("{T}_GenericOf_{0}")]`). Will mull on the problem of languages with overloads/templates a little more.

**Comment by Manishearth** [comment]
A problem with this is that it requires everything to be in the same module.

How about instead something like

```rust
#[diplomat::genericset(T: SomeOpaqueType, SomeOtherOpaqueType, u8)]
pub struct GenericOpaque<T: SomeTrait>(RustType<T>);

impl GenericOpaque {
   fn create(t: &T) -> Box<Self> {...}
}
```

The `genericset` argument can also go on generic methods defined in another module.

**Comment by tyler-zeromatter** [comment]
I think that could work, although I'd ideally like something closer to native Rust syntax so that it makes sense for developers. Could we just repurpose `#[diplomat::include("some_file.rs")]` to handle trait definitions as well?

```rs
// some_file.rs
#[diplomat::generictrait]
pub trait SomeTrait {...}

impl SomeTrait for u8 {}

// include_trait.rs
#[diplomat::include("some_file.rs")]
#[diplomat::bridge]
mod ffi {
  pub struct GenericStruct<T: SomeTrait> {}
  // Might even be possible to parse additional trait impls after definition in this file (although maybe we'd want to error on this if not included directly after a trait definition? Since we parse in order):
  impl SomeTrait for u16 {}
}
```

**Comment by Manishearth** [comment]
Yes, we could! I'm worried about `include` needing to get used so much that it slows things down (remember, this affects macro build times) but maybe it's fine.

**Comment by tyler-zeromatter** [comment]
You mean that it would take longer to parse `#[diplomat::include]` in the proc macro because of the need to parse more `syn` syntax in the other file?

**Comment by Manishearth** [comment]
Yes. This is already a problem with macros but I'm worried it would get worse with other things. Worth measuring: you guys are the ones actually using macros.

**Comment by tyler-zeromatter** [comment]
Currently Diplomat has fairly limited support of Rust's macro features, so internally we're not using a whole lot of macros (I count around 7 that we use `#[diplomat::include]` for), and even then because we currently don't do macro recursion the footprint of each is small. Currently running some timings, so will get back to you on that

**Comment by tyler-zeromatter** [comment]
We don't have a lot of definitions, but because there's no caching in the macro, parsing the `syn` info of each `#[diplomat::include]` and adding it to the module takes around 4-6ms. That's around 55ms total of parsing `#[diplomat::include]` statements for the whole project. It could get bigger if we ever implement more macro features like recursion.

I think it should still remain pretty small for traits, if the syntax is limited to defining an empty trait and what types it's implemented for; not sure if we're interested in expanding the syntax for more features?

**Comment by tyler-zeromatter** [comment]
Spun up a separate discussion for `#[diplomat::include]` performance concerns: https://github.com/rust-diplomat/diplomat/discussions/1245

**Comment by Manishearth** [comment]
Sounds good!

My other concern with traits is that this is where "real" code would have to mix with diplomat "marker" code (people are going to want to use real traits here so we'll have to maybe inherit Trait A from Trait B or something) and that can get messy in Diplomat. We try to avoid the macro ever having t odeal with "real" code, it only looks at APIs that are more like an IDL.

**Comment by tyler-zeromatter** [comment]
I have a lot more I'd like to discuss here, but I think things within Diplomat are super busy this week so I'll table this and come back with some more thoughts on a later date

**Comment by Manishearth** [comment]
Yep! I don't mind continuing this discussion though. The 1.0 discussion isn't going to resolve soon, I think.
